ambry-store/src/test/java/com.github.ambry.store/IndexTest.java

```diff
@@ -697,18 +702,89 @@ public void findEntriesSinceOnRestartTest() throws IOException, StoreException {
+    // it will be the starting offset of last entry which is firstRecordFileSpan.getStartOffset())
+    expectedEndToken = new StoreFindToken(firstRecordFileSpan.getStartOffset(), sessionId, incarnationId, false);
+    expectedEndToken.setBytesRead(bytesRead);
+    doFindEntriesSinceTest(startToken, Long.MAX_VALUE, allKeys.keySet(), expectedEndToken);
+
+    // create a token that is not past the index end offset on startup after recovery, but with old incarnationId.
+    // again the token should be reset internally, all keys should be returned and the returned token should be correct
+    // (offset in it will be the starting offset of last entry which is firstRecordFileSpan.getStartOffset())
+    startToken = new StoreFindToken(lastRecordOffset, oldSessionId, oldIncarnationId, false);
+    expectedEndToken = new StoreFindToken(firstRecordFileSpan.getStartOffset(), sessionId, incarnationId, false);
+    expectedEndToken.setBytesRead(bytesRead);
+    doFindEntriesSinceTest(startToken, Long.MAX_VALUE, allKeys.keySet(), expectedEndToken);
+
+    // test cases where no new puts/deletes have gone in after restart.
+    // after restart, replica 1(assuming no new puts have gone in) will return a journal based token referring to
+    // logEndOffsetOnStartUp to replica2.
+    // On the subsequent findEntriesSince() call, the index should start returning entries starting from that offset
+    // and should not consider it as non-inclusive offset
+    recovery = new MessageStoreRecovery() {
+      @Override
+      public List<MessageInfo> recover(Read read, long startOffset, long endOffset, StoreKeyFactory factory)
+          throws IOException {
+        return Collections.emptyList();
+      }
+    };
+    reloadIndex(true);
+    lastRecordOffset = index.journal.getLastOffset();
+    bytesRead = log.getDifference(firstRecordFileSpan.getEndOffset(), new Offset(log.getFirstSegment().getName(), 0));
+    // create a token that will be past the index end offset on start up after restart.
+    startToken = new StoreFindToken(
+        new Offset(log.getEndOffset().getName(), (lastRecordOffset.getOffset() + 2 * PUT_RECORD_SIZE)), oldSessionId,
+        incarnationId, false);
+    // end token should point to offset of last entry i.e. firstRecordFileSpan.getEndOffset()
+    expectedEndToken = new StoreFindToken(firstRecordFileSpan.getEndOffset(), sessionId, incarnationId, true);
+    expectedEndToken.setBytesRead(bytesRead);
+    // Fetch the FindToken returned from findEntriesSince
+    FindInfo findInfo = index.findEntriesSince(startToken, Long.MAX_VALUE);
+    Assert.assertEquals("EndToken mismatch ", expectedEndToken, findInfo.getFindToken());
+    Assert.assertEquals("No entries should have been returned ", 0, findInfo.getMessageEntries().size());
+
+    // add 2 new entries to the log
+    Offset endOffsetOfPrevMsg = index.getCurrentEndOffset();
+    appendToLog(PUT_RECORD_SIZE);
```

Comment:
You could use the start offset of the returned `FileSpan` to look up the `StoreKey` in the journal.